### ISSUE TITLE
Home Screen - tasks extensibility

### ISSUE THREAD

**Issue by pmcpinto** [woocommerce/woocommerce-admin#5242]
**User story:**
As a 3PD
I want to easily extend the home screen task feature
So I can incorporate recurrent actions in my extension that require immediate attention

**Description:** our goal is to allow 3PDs to create home screen tasks for their extensions, in order to have a cohesive user experience. Ideally we should allow them to implement two different types of tasks:
- Setup tasks: that will be incorporated in the setup task list
- Recurring tasks: following the same approach that we have for orders, reviews and stock

We have a [wc-admin task example](https://github.com/woocommerce/woocommerce-admin/blob/master/docs/examples/extensions/add-task/woocommerce-admin-add-task-example.php), but I'm not sure if it can be reused for this.

**Acceptance criteria:**
- [ ] Allow 3pds to create setup tasks for their extensions with the following items: task name (with link to an extension page), time to complete the tasks and dismiss feature
- [ ] If a merchant completes the task list and then installs and extension that has a setup task associated, then we should display the task list again with that extension
![Home new extension.png](https://images.zenhubusercontent.com/5d84fe8e1201de0001b6ab61/1ea9cd21-5a6d-4f6b-b1ad-8d8d062adca5)
- [ ] Allow 3pds to create recurring tasks and keep the limit that we have for our tasks: max 5 items per task

**Event tracking:**
NA

**Additional links:**
[P2 with links to the Figma file](pbIJXs-ma-p2)

**Comment by samueljseay** [comment]
I think this is possibly several tasks and I'd like to break it up further, but first I want to open a discussion about a possible approach for this.

Right now we have setup management tasks (and checks for if they're complete) defined client side. We don't have a good story for doing client side integration with WooCommerce Admin right now, so it makes sense to move the setup task list to PHP.

I'm proposing a task to move the existing setup management into a server side API and remove that from the client. This would be a good opportunity to simplify the client side including being able to query if the task list is complete from the API, instead of our current rather flaky client side code that checks list completion status.

Once we create an API we can use to query setup tasks, it will be a simpler next step to introduce the creating of new setup tasks via PHP. We won't have to merge our existing client side implementation with the backend one which will likely lead to more subtle issues and bugs imho.

@jeffstieler @octaedro keen to hear what you think of this approach.

One question for @pmcpinto , how relevant is tracking of completing the entire task list after these features are completed? It feels like the task list will be very open ended, is that tracking event still relevant?

**Comment by pmcpinto** [comment]
> One question for @pmcpinto , how relevant is tracking of completing the entire task list after these features are completed? It feels like the task list will be very open ended, is that tracking event still relevant?

@samueljseay are you talking about this one: `wcadmin_tasklist_completed`?

**Comment by samueljseay** [comment]
@pmcpinto Looking at it, I that event was for entire completion of the task list but it looks like it is an event per action on the list, so thats my mistake, I think we're ok not to change anything there. One more question then:

do we want to track the event completion performed on 3pd tasks?

**Comment by samueljseay** [comment]
After some thought I think this feature is more complex than it appears on the surface and to do it we'll need to break up a few sub tasks to do the work. I think it's worth mentioning that those subtasks will represent a significant amount of effort. @pmcpinto  if there are several new tasks with significant estimates would that make you want to deprioritise this over other features for this cycle? (I don't have an official estimate but I have high confidence these tasks could swallow most or all of the cycle).

There are some reasons why this is more complex than it seems:

Our current internal solution for task management is brittle (and error prone as we've seen in a lot of our recent bugs). it cannot be extended by a 3PD in its current state. Virtually all the logic for it resides in JavaScript. Specifically the logic around when a task is completed is something that happens in JavaScript.

We don't have good prior art for JavaScript integration right now. We do offer some currently such as in reports, but it's hard to use and requires a lot of upfront coding from 3DPs. Even if we were to somehow *shoehorn* this into the client, this would be a bad experience for 3DPs who are wanting to do something fairly trivial from their perspective.

If we have broader conversations about how to integrate things via JavaScript I think we will end up derailing into long winded discussions because it's not easy to solve. It's a lot more complex to do well. The more complex it is to use, the less 3DPs will use it. Many 3PDs are not familiar with JavaScript or React.

To fix the problem it starts by fixing our own problem: we need to refactor the task completion to be handled in PHP and expose a REST API we can consume and update for task status, details, and completion. Once this new API exists we can remove our brittle logic from the client side and we can get the tasks from this REST API ourselves and we will get to "dogfood" our own integration to ensure it is easy to use.

The second step will be some PHP APIs to perform task updates such as completion or other detail changes. This will make the 3PD's task of updating status of their setup item much simpler.

With those 2 parts in place we'll get some benefits:
1. Adding new setup tasks of our own in future will be a simple straightforward path, and like the example of inbox notifications we may even get opportunities to offload adding new setup tasks on to other teams who may have interest.
2. By having to use our own APIs we'll ensure they work well for other developers, we'll catch bugs or pain points and be able to fix them.
TLDR; This feature is complex and I see several new tasks arising to solve it:
1. Develop a REST API for task completion
2. Remove our current task completion logic in JS and use this new API
3. Develop a PHP API for task completion including a filter to add tasks to the list and objects for managing the task details. 

cc @jeffstieler

**Comment by pmcpinto** [comment]
@samueljseay here's my feedback:

> @pmcpinto Looking at it, I thought that event was for entire completion of the task list but it looks like it is an event per action on the list, so thats my mistake, I think we're ok not to change anything there. One more question then:
> do we want to track the event for task completion performed on 3pd tasks?

Yes, it would be good - mainly to have a sense of how many extensions are taking advantage of the extensibility. 

> After some thought I think this feature is more complex than it appears on the surface and to do it we'll need to break up a few sub tasks to do the work. I think it's worth mentioning that those subtasks will represent a significant amount of effort. @pmcpinto if there are several new tasks with significant estimates would that make you want to deprioritise this over other features for this cycle? (I don't have an official estimate but I have high confidence these tasks could swallow most or all of the cycle).

Thanks for outlining the complexity behind adding extensibility. Given the potential amount of effort required to implement extensibility, I think that we should backlog this and evaluate implementing it in the upcoming cycle. Shipping the activity panels in the home screen first and then work on extensibility allows us to have a sense of the 3PDs appetite to take advantage of this feature and get feedback about potential use cases that we're still not covering. Just two examples, we already received feedback from merchants saying that they'd like to use tasks as internal reminders to complete specific actions and store builders saying that it would be useful to create subsequent tasks for merchants, to facilitate the hand-off between builder and merchants. 

What do y'all think?

**Comment by samueljseay** [comment]
That sounds good to me. Thanks @pmcpinto

**Comment by samueljseay** [comment]
@pmcpinto 🤦 I have to go back on my word on this one after some further exploration. I thought after some initial searches that the original integration had been removed but it turns out I was wrong. I don't know how I missed it, but we **do** have this functionality already as a client side integration to the task list. In fact the current integration does almost everything we requested here already.

What we have is a JS filter: `woocommerce_admin_onboarding_task_list`. And I'll try answer what it does currently against the requirements here.

> Allow 3pds to create setup tasks for their extensions with the following items: task name (with link to an extension page), time to complete the tasks and dismiss feature

This is pretty much all supported right now but it doesn't link to an extension page. They can render whatever they would like to display in the actual tasks to perform. task name, time to complete are supported. They could dismiss their task from the "items" part, but there is not support for dismissing the task from the task list currently.

> If a merchant completes the task list and then installs and extension that has a setup task associated, then we should display the task list again with that extension

This is one part that is not supported at the moment. Once the task list has been completed an initial time it doesn't come back. The option we set to "complete" it is `woocommerce_task_list_complete`. We will need to change this approach.

>  Allow 3pds to create recurring tasks and keep the limit that we have for our tasks: max 5 items per task

There is no API to create a recurring task, but they could do it "somehow" by coordinating their plugin with the JS they write to add the task. Does it need to be more full featured than this at the moment? One way I can imagine they could do this is with a WP cron job and setting some data via options or something that they could then query when displaying their tasks in JS.

As for the "5 items per task" there is no way currently to enforce a limit. They get to render whatever they like for the task part, I don't think we could limit it without changing the way the integration works significantly.

So I think if we stick with the existing task list we have these things to implement (and with some notes for dev implementing)

1. Add support for dismissing a task from the task list (What would happen here? maybe we could have an onDismiss callback where they can handle the task themselves? We can't force it to go away under the current implementation since its just a filter).

2. Ensure that every time the task list is rendered it will show if there are any incompleted tasks remaining, even if it was completed in the past.

cc @jeffstieler @octaedro

**Comment by pmcpinto** [comment]
@samueljseay thanks for the clarification. I have a couple of questions:

> This is pretty much all supported right now but it doesn't link to an extension page. They can render whatever they would like to display in the actual tasks to perform. task name, time to complete are supported. They could dismiss their task from the "items" part, but there is not support for dismissing the task from the task list currently.

If I understand correctly, they can redirect the setup task to any place they want, right? (a link to a specific page in their extension, render a new page exclusive for the setup as we're doing in our tasks, etc)

> There is no API to create a recurring task, but they could do it "somehow" by coordinating their plugin with the JS they write to add the task. Does it need to be more full featured than this at the moment? One way I can imagine they could do this is with a WP cron job and setting some data via options or something that they could then query when displaying their tasks in JS.

Essentially a 3PD could take advantage of `woocommerce_admin_onboarding_task_list` to create recurrent tasks as well, correct? If yes, then the tasks would always show in the setup task list or maybe I'm missing something here

**Comment by samueljseay** [comment]
> If I understand correctly, they can redirect the setup task to any place they want, right? (a link to a specific page in their extension, render a new page exclusive for the setup as we're doing in our tasks, etc)

The way it works currently is we always provide a "space" to render the task. it always goes to the same place and displays that "space". Its very "janky" but it is "possible" for them to redirect when we render that "space". 

They may not feel like its intended to work that way (if that makes sense). The way it would work is they would render a blank setup task and as it renders they could redirect elsewhere). So I don't think that experience is very nice, but it could be a workaround.

> Essentially a 3PD could take advantage of woocommerce_admin_onboarding_task_list to create recurrent tasks as well, correct? If yes, then the tasks would always show in the setup task list or maybe I'm missing something here

Yes, we haven't provided any extra way for them to say "heres a task I want to display every `n` days, but they certainly could do it. I hope that explains it well, we're venturing into the realm of "developer experience" 😄

**Comment by pmcpinto** [comment]
Thanks for the clarification

> Yes, we haven't provided any extra way for them to say "heres a task I want to display every n days, but they certainly could do it. I hope that explains it well, we're venturing into the realm of "developer experience" 😄

For recurring tasks, I was seeing as allowing them to take advantage of the feature experience that we're building for orders, inventory and reviews tasks. But probably it will be better to ship them first and then make it extensible.

**Comment by samueljseay** [comment]
I'm just going to summarise the discussion here for whoever picks this up:

1. We have some extensibility already that is almost fit for purpose. 
2. We won't implement recurring tasks as a first class citizen here, we can look to do that later (I think thats right @pmcpinto ? )

There are 2 things to implement:

- [ ] Add support for dismissing a task from the task list (I'm suggesting we provide an `onDismiss` callback where they can handle the task themselves. 

**Note: We can't force it to go away because the implementation of adding tasks is just a filter, 3PD will need to handle this.**

- [ ] Ensure that every time the task list is rendered it will show if there are any incompleted tasks remaining, even if it was completed in the past. 

One question for clarification @pmcpinto I assume we will still not show the task list again if it has been **dismissed** in the past? I can't think of a really reliable way to handle that so I would say thats another limitation at this stage.

**Comment by pmcpinto** [comment]
> We won't implement recurring tasks as a first class citizen here, we can look to do that later (I think thats right @pmcpinto ? )

Yep, that's correct

> One question for clarification @pmcpinto I assume we will still not show the task list again if it has been dismissed in the past?

That would be the ideal state. We if a user dismisses the "Core" task list without completing all the steps and then a 3PD triggers a new task it will display the 3PD task + the "Core" tasks, correct?

**Comment by samueljseay** [comment]
@pmcpinto and I discussed this further in DM so I'll try summarize those thoughts here along with my current thinking.

Summary:

The way extensibility works right now is via a "filter". Filters just append items to a list, so there is no good way to synchronize it with whether the task list was dismissed. Imagine a scenario where the user just clicked "dismiss" of the task list, but there are still a bunch of 3PD tasks in the list, do we display it or not? I don't think its easy to determine which happened first, the dismiss or the add of the 3PD task.

@pmcpinto and I discussed the possibility of some new logic to have a separate list for 3PD tasks. the logic would work something like (quoted from the discussion): 

"I’m wondering if it would make sense to implement a logic like:
If core task list enabled -> display 3PDs tasks within the task list
If core task list completed/dismissed -> display each 3PD tasks as an individual task with a dismiss button"

@elizaan36 keen to get some feedback from you on this issue. Does that logic sound ok? Is there additional design work required for that approach? (I think it would be helpful to make sure we cover all the UI states that might arise to just help the dev implementing that).

I'll label this as blocked for now, until we can make a decision on this.

**Comment by elizaan36** [comment]
> If core task list enabled -> display 3PDs tasks within the task list

Yep, this is aligned with the existing design. Any tasks created by extensions before the task list is dismissed would appear in the list. For example, in this example a merchant has completed the Abandoned cart task in the list.

![Home - Setup tasks](https://user-images.githubusercontent.com/5121465/99609012-ac923280-29d4-11eb-8dad-ac0a0a67252b.png)

> If core task list completed/dismissed -> display each 3PD tasks as an individual task with a dismiss button"

It seems reasonable to generate a new card for any tasks added after the original was dismissed, mainly because it wouldn't make sense to show the dismissed task list again with all the completed tasks alongside the new one.

I'm not sure the design needs to change that much, since any 3PD extensions could generate a new task list using the same card component. The title should be tweaked to differentiate it from the original task list, maybe `To do's`.

![Home - To do's](https://user-images.githubusercontent.com/5121465/99609944-735ac200-29d6-11eb-92c7-ceb0433eadfe.png)

The only thing that might be problematic is if each 3PD created their own card for each separate task. Could we use the filter to append all new 3PD items to this `To do's` card once the initial setup task list is dismissed?

**Comment by pmcpinto** [comment]
> I'm not sure the design needs to change that much, since any 3PD extensions could generate a new task list using the same card component. The title should be tweaked to differentiate it from the original task list, maybe To do's.

I'm not sure if "To do's" is the best option. It can be a concept that we can implement in the future to allow users to set up reminders, etc.

**Comment by elizaan36** [comment]
Maybe creating a separate component for all extension tasks is the answer then, regardless of whether the task list has been dismissed. I think it's a little confusing (and makes the logic more complex) when extension tasks are sometimes in the initial setup list and sometimes in another place.

**Comment by samueljseay** [comment]
> Maybe creating a separate component for all extension tasks is the answer then, regardless of whether the task list has been dismissed. I think it's a little confusing (and makes the logic more complex) when extension tasks are sometimes in the initial setup list and sometimes in another place.

I agree a separate component would be nice. @elizaan36 just to confirm, for a separate component would we need any more design work for that or would it just be another panel with a different header? Keen to know the copy for that if possible.

It sounds like thats the last missing piece to continue with this bit of work.

**Comment by pmcpinto** [comment]
> Maybe creating a separate component for all extension tasks is the answer then, regardless of whether the task list has been dismissed. I think it's a little confusing (and makes the logic more complex) when extension tasks are sometimes in the initial setup list and sometimes in another place.

👍 seems to be the best option. So we would have the store setup list and then another that would be called "extensions setup" or something like that, correct? I think it can be fairly intuitive if we nail down the copy

**Comment by elizaan36** [comment]
> @elizaan36 just to confirm, for a separate component would we need any more design work for that or would it just be another panel with a different header?

I'd like to visually differentiate it in some way, but might need a minute to do that (can share something next week.) 

In the meantime, we can just use the same card component with the different header. Here's what that could look like.
![Home - Setup tasks extensibility](https://user-images.githubusercontent.com/5121465/100138883-43874080-2e54-11eb-9e61-15229f48213a.png)

A couple notes on this:
- Going this route would mean that the "Set up WooCommerce Payments" task would appear in the "Extensions setup" card.
- Also wanted to confirm that both task lists appear below the activity panel by default if the user has orders/stock/reviews.

**Comment by samueljseay** [comment]
Looks great to me @elizaan36 thank you!

**Comment by pmcpinto** [comment]
@samueljseay I would be great to create a tutorial for the Woo dev blog once this work is done

**Comment by samueljseay** [comment]
Ok thanks @pmcpinto will do.

**Comment by budzanowski** [comment]
I have an extensibility question. I am looking at adding Storefront related setup tasks to the Task List. What I found is that `Personalize my store` largely overlaps with Storefront internal setup. Is there a possibility to modify the behavior of already existing setup steps? For example, adding a step into `Personalize my store` lists that would be used by Storefront to do any necessary actions, or replace the homepage setup step with the Storefront homepage setup.

With the current architecture, I don't see this as possible. But it looks more doable when everything would be moved to PHP with JS only consuming task list items. Do I understand this correctly?

**Comment by octaedro** [comment]
Hi @budzanowski,

> Is there a possibility to modify the behavior of already existing setup steps? For example, adding a step into Personalize my store lists that would be used by Storefront to do any necessary actions, or replace the homepage setup step with the Storefront homepage setup.

Unfortunately, it's not possible to extend the behavior of the already existing setup steps without modifying the code. If we'd want to have the described behavior we should modify [this component](https://github.com/woocommerce/woocommerce-admin/blob/main/client/task-list/tasks/appearance.js) directly.

> With the current architecture, I don't see this as possible. But it looks more doable when everything would be moved to PHP with JS only consuming task list items. Do I understand this correctly?

Currently, we have a blended option (PHP + Javascript) that won't change for now. The [PR 5794](https://github.com/woocommerce/woocommerce-admin/pull/5794) enhances the extensibility options but AFAIK a bigger refactor is not in the scope.
